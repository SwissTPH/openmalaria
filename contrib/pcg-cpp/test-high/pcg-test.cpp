/*
 * PCG Random Number Generation for C++
 *
 * Copyright 2014-2017 Melissa O'Neill <oneill@pcg-random.org>,
 *                     and the PCG Project contributors.
 *
 * SPDX-License-Identifier: (Apache-2.0 OR MIT)
 *
 * Licensed under the Apache License, Version 2.0 (provided in
 * LICENSE-APACHE.txt and at http://www.apache.org/licenses/LICENSE-2.0)
 * or under the MIT license (provided in LICENSE-MIT.txt and at
 * http://opensource.org/licenses/MIT), at your option. This file may not
 * be copied, modified, or distributed except according to those terms.
 *
 * Distributed on an "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, either
 * express or implied.  See your chosen license for details.
 *
 * For additional information about the PCG random number generation scheme,
 * visit http://www.pcg-random.org/.
 */

/*
 * This file is based on the demo program for the C generation schemes.
 * It shows some basic generation tasks.
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <cassert>
#include <climits>
#include <iostream>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <numeric>
#include <random>       // for random_device

#include "pcg_random.hpp"

// This code can be compiled with the preprocessor symbol RNG set to the
// PCG generator you'd like to test.

#ifndef RNG
    #define RNG pcg32
    #define TWO_ARG_INIT 1
#endif

#define STRINGIFY_IMPL(x) #x
#define STRINGIFY(x)      STRINGIFY_IMPL(x)

using namespace std;
using pcg_extras::operator<<;

#if !PCG_EMULATED_128BIT_MATH || !AWKWARD_128BIT_CODE

int main(int argc, char** argv)
{
    // Read command-line options

    int rounds = 5;
    bool nondeterministic_seed = false;

    ++argv;
    --argc;
    if (argc > 0 && strcmp(argv[0], "-r") == 0) {
        nondeterministic_seed = true;
        ++argv;
        --argc;
    }
    if (argc > 0) {
        rounds = atoi(argv[0]);
    }

    /* Many of the generators can be initialized with two arguments; the second
     * one specifies the stream.
     */

#if TWO_ARG_INIT
    RNG rng(42u, 54u);
#else
    RNG rng(42u);
#endif

    if (nondeterministic_seed) {
        // Seed with external entropy from std::random_device (a handy
        // utility provided by pcg_extras).
        rng.seed(pcg_extras::seed_seq_from<std::random_device>());
    }

    constexpr auto bits = sizeof(RNG::result_type) * CHAR_BIT;
    constexpr int how_many_nums = bits <= 8  ? 14
                                : bits <= 16 ? 10
                                :              6;
    constexpr int wrap_nums_at  = bits >  64 ? 2
                                : bits >  32 ? 3
                                :              how_many_nums;

    cout << STRINGIFY(RNG) << ":\n"
    //   << "      -  aka:         " << pcg_extras::printable_typename<RNG>()
    // ^-- we skip this line because the output is long, scary, ugly, and
    //     and varies depending on the platform
         << "      -  result:      " <<  bits << "-bit unsigned int\n"
         << "      -  period:      2^" << RNG::period_pow2();
    if (RNG::streams_pow2() > 0)
         cout << "   (* 2^" << RNG::streams_pow2() << " streams)";
    cout << "\n      -  size:        " << sizeof(RNG) << " bytes\n\n";

    for (int round = 1; round <= rounds; ++round) {
        printf("Round %d:\n", round);

        /* Make some N-bit numbers */
        cout << setw(4) << setfill(' ') << bits << "bit:";
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << endl;

        cout << "  Again:";
        rng.backstep(6);
        for (int i = 0; i < how_many_nums; ++i) {
            if (i > 0 && i % wrap_nums_at == 0)
                cout << "\n\t";
            cout << " 0x" << hex << setfill('0') 
                 << setw(sizeof(RNG::result_type)*2) << rng();
        }
        cout << dec << endl;

        /* Toss some coins */
        cout << "  Coins: ";
        for (int i = 0; i < 65; ++i)
            cout << (rng(2) ? "H" : "T");
        cout << endl;

        RNG rng_copy{rng};
        /* Roll some dice */
        printf("  Rolls:");
        for (int i = 0; i < 33; ++i)
            cout << " " << (uint32_t(rng(6)) + 1);
        cout << "\n   -->   rolling dice used "
             << static_cast<size_t>(rng - rng_copy)
             << " random numbers" << endl;

        /* Deal some cards using pcg_extras::shuffle, which follows
         * the algorithm for shuffling that most programmers would expect.
         * (It's unspecified how std::shuffle works.)
         */      
        enum { SUITS = 4, NUMBERS = 13, CARDS = 52 };
        char cards[CARDS];
        iota(begin(cards), end(cards), 0);
        pcg_extras::shuffle(begin(cards), end(cards), rng);

        /* Output the shuffled deck */
        printf("  Cards:");
        static const signed char number[] = {'A', '2', '3', '4', '5', '6', '7',
                                             '8', '9', 'T', 'J', 'Q', 'K'};
        static const signed char suit[] =   {'h', 'c', 'd', 's'};
        int i = 0;
        for (auto card : cards) {
            ++i;
            cout << " " << number[card / SUITS] << suit[card % SUITS];
            if (i % 22 == 0)
                cout << "\n\t";
        }
        
        cout << "\n" << endl;
    }

    // Ensure that the input/output functions work.
    std::stringstream ss;
    ss << rng;
    RNG from_string;
    ss >> from_string;
    assert(from_string == rng);

    return 0;
}

#else //  i.e. PCG_EMULATED_128BIT_MATH && AWKWARD_128BIT_CODE

int main()
{
    // It's not that it *can't* be done, it just requires either C++14-style
    // constexpr or some things not to be declared const.
    cout << "Sorry, " STRINGIFY(RNG) " not supported with emulated 128-bit math"
         << endl;
}

#endif
