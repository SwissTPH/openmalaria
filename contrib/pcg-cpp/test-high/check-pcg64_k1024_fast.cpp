#define RNG pcg64_k1024_fast
#define TWO_ARG_INIT 0
#define AWKWARD_128BIT_CODE 1

#include "pcg-test.cpp"

