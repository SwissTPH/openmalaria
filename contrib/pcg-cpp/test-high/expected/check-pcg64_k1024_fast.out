pcg64_k1024_fast:
      -  result:      64-bit unsigned int
      -  period:      2^65664
      -  size:        8208 bytes

Round 1:
  64bit: 0x716e1291ecd614bf 0x6d9c60c68bfa5bac 0x6011e6dee28a425f
	 0x3a93df63ad5d8c2d 0x8c276f08b088c497 0x4d7204001d16777d
  Again: 0x716e1291ecd614bf 0x6d9c60c68bfa5bac 0x6011e6dee28a425f
	 0x3a93df63ad5d8c2d 0x8c276f08b088c497 0x4d7204001d16777d
  Coins: HHHHHHHHTTHTTTTTTHTTTTHHHHTTTHTHTHHTTTHTHTTHHHTTHHHTTTTHTHHHTHHHT
  Rolls: 6 3 1 2 2 3 3 2 5 6 1 5 2 1 1 3 1 4 5 2 4 5 3 4 2 2 5 5 3 1 2 6 4
   -->   rolling dice used 33 random numbers
  Cards: Ts 2h 3c 9d 3s 2c Th 3h 4c Qh Js Ad 7c 9c Td 8s 7h 4s 6c 6h Tc 7d
	 8d Jc Qs Ac Jd 5s 6d 4d Kc Kd Jh 8h Qd 2s 5d Kh 3d 4h 8c Qc 9h 6s
	 5c As Ah Ks 7s 5h 9s 2d

Round 2:
  64bit: 0xbd047092dbc12c12 0xb8023964c5590845 0x6a2acc287da9b782
	 0xf27a18fdeaca790a 0x5dd53aa3af4dd73e 0x407a59d050593064
  Again: 0xbd047092dbc12c12 0xb8023964c5590845 0x6a2acc287da9b782
	 0xf27a18fdeaca790a 0x5dd53aa3af4dd73e 0x407a59d050593064
  Coins: TTHTHTTTHHHTTHTTTHHTHHTTTHHTHHHHHTHTHTTHTTHHHTTTTTTTTTHTHTHTHTTTT
  Rolls: 6 2 6 6 4 4 4 4 4 1 3 6 1 2 5 4 4 6 2 5 3 5 5 6 6 2 3 5 3 6 2 4 5
   -->   rolling dice used 33 random numbers
  Cards: 4c 4d 9c 6d Qs 3d 5c Tc 9d Ts 2h Th Qh Jh 6h 8s 2s Ac 3h 7s 4s 5h
	 9h Js Jc 2c 4h Kd Ah 5d 8c As 5s Kc Qd 7c Ks Ad Qc 8h 6c 7d Jd 8d
	 2d 3c 7h Kh 6s 3s Td 9s

Round 3:
  64bit: 0xb9f6e508611fa39b 0x50ff50a7fda1c6ff 0xc9fc4c468fc0e452
	 0x13992edc8b596e39 0x6e621625efd5f7df 0xe0ec108a27c75fae
  Again: 0xb9f6e508611fa39b 0x50ff50a7fda1c6ff 0xc9fc4c468fc0e452
	 0x13992edc8b596e39 0x6e621625efd5f7df 0xe0ec108a27c75fae
  Coins: HTHTTTHTHTHTTHTTHHTTHTHHHHTTTHHHTHHHHTHHHTHHTTHTTHTHHTHTTHHHTHTHT
  Rolls: 4 5 2 2 5 6 4 5 6 3 6 5 6 6 4 5 1 5 5 1 4 4 6 6 2 2 3 5 1 4 3 3 4
   -->   rolling dice used 33 random numbers
  Cards: 3d 5h 2d 7s Ts Tc 3s Kd 8c 6c 4s Js 9s Jd 4d Qc 2s As 6d Jc Qs Ad
	 8h Kc Td 7h 7c 4h 6h 2h 5c 9d 4c 2c Qh Jh Ks 3c 9h 8d Ah Ac Qd Th
	 7d 3h 8s 5s 9c 6s Kh 5d

Round 4:
  64bit: 0xb5cb75ddd90a0f9e 0x4cf347f82cce20c6 0x6243c21a2332efc2
	 0x43f3e96d96ce7569 0xea93a4709c0f83f8 0x88c99d75a1dec4a9
  Again: 0xb5cb75ddd90a0f9e 0x4cf347f82cce20c6 0x6243c21a2332efc2
	 0x43f3e96d96ce7569 0xea93a4709c0f83f8 0x88c99d75a1dec4a9
  Coins: HTTHTHHHTTTHTTTTHHTHTHHTTTHHHTHHHHTTTTHHHTHHTHTHHTTTTTHHTHHTTHTHT
  Rolls: 5 3 1 3 6 3 5 3 3 5 6 1 5 5 3 1 2 1 4 5 2 3 5 4 3 6 1 6 4 3 5 3 1
   -->   rolling dice used 33 random numbers
  Cards: 7d 6c 4d 8s Td Kd Qh 9h 9c 5d 7s Tc Qs Ks 9s 6d As Jc 8h Qd 4s 4c
	 Kc 7h 7c Jd 4h 2c 8c 2d 5s 2s 3d Jh Kh 6h 2h Ts Ac Js 9d Ad 8d Qc
	 Th 3c 6s 3s 5c Ah 3h 5h

Round 5:
  64bit: 0xa9cba0879f523763 0x9d063cc605508ab9 0xaa9dd9f6f8792bdc
	 0xa3005a22d15b695b 0x66f3b5676edeceea 0x8e315e033cded525
  Again: 0xa9cba0879f523763 0x9d063cc605508ab9 0xaa9dd9f6f8792bdc
	 0xa3005a22d15b695b 0x66f3b5676edeceea 0x8e315e033cded525
  Coins: THTTHTHHHHTTHTTHHHHHHHTTHTHTHTTTTHHHTHHHTTHTTHTTHHHTHHHHTHTTTTTHT
  Rolls: 4 4 1 4 6 2 3 1 6 5 5 2 2 6 3 1 6 2 3 2 2 1 6 1 4 1 1 5 2 6 4 5 2
   -->   rolling dice used 33 random numbers
  Cards: Th 2c 2s Td 4s Tc Jh Js 6h 2h Ac Qs 9h Kd 9s Qh 6d 6s Ad 2d 9c 3h
	 7h 8s Qc 5d 3d 5h Ah Jd 3s As 9d 5c 6c Ks 7c Jc Qd 4c 5s 8c 3c Ts
	 Kc 8h 7s 4h Kh 7d 4d 8d

